### PR TITLE
Use streaming parser

### LINKED ISSUE #312: opencryptoki.conf parser mismatch in migration tool
In pkcstok_migrate.c we have the lines
`    snprintf(parm, sizeof(parm), "slot %ld", slot_id);
    while (fgets(line, sizeof(line), fp)) {
        if (!begins_with(line, '#') && strstr(line, parm)) {
            ret = get_stdll_name(fp, dll_name);
            goto done;
        }
    }
`
This parser is in no way aligned to the parser of the slot daemon.  Furthermore, it wrongly parses a lot of configurations:

- "slot  1" (with two spaces after the "slot" keyword)
- searching for slot 1 might hit slot 10 (since we do a prefix search)
- if the line does not contain a "{" after the slot number, the config line does not get rejected
- 80 spaces after "slot" can never be parsed with this simple solution, but according to the grammar of the slot daemon, this would be valid.

We should align these parsers.

### PR DESCRIPTION
**Pull request by juergenchrist**
Restructured the parser as a streaming parser that emits events to an external
processing entity.  With this method we can use the parser also in the
migration tool to find the correct stdll and also update the configuration
file while preserving comments and newlines (tabs and spaces are not
necessarily preserved).

The rework of the parser also updated the grammar to include empty lines and
whole line comments within the slot definition.  This resolves the ugliness

```
slot 1 {
 description = something # comment allowed here
}
```

versus

```
slot 1 {
# comment not allowed here
  description = something
}
```
Fixes #312 

Signed-off-by: Juergen Christ <jchrist@linux.ibm.com>

### CODE REVIEW

**Comment by ifranzki** [comment]
Looks pretty nice! I was expecting/fearing a more complex change, but this seems to be pretty straight forward. 

One minor thing: 
lexer.l seems to have bene moved with 'git mv', so git still knows where the file was located previously.
However, for parser.y this does not seem to be the case? 
```
---
 usr/lib/common/configparser.h               |  63 +++++
 usr/{sbin/pkcsslotd => lib/common}/lexer.l  |  15 +-
 usr/lib/common/parser.y                     | 217 +++++++++++++++
 usr/sbin/pkcsslotd/parser.y                 | 289 --------------------
```

**Comment by juergenchrist** [comment]
Weird thing is that neither file has been moved via git mv.  Will try to get this as proper tracking of the move.

**Comment by juergenchrist** [comment]
Ups, forgot about the find-renames heuristic.  Since the files changed a lot, this heuristic build into git does not detect them with the default threshold of 50% (more than 50% of the file stay the same => rename; otherwise add/delete pair).  A quick check shows that lexer.l has a 93% similarity, parser.y only a 34%.  So try the option `--find-renames=34` to get this as a rename.

File: usr/lib/common/parser.y [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+	conf = fopen(configfile, "r");
+
+	if (!conf) {
+		fprintf(stderr, "Failed to open %s: %s\n", configfile, strerror(errno));
+		return -1;
+	}
+
+	yyin = conf;
+    parsefuncs = funcs;
```
Comment (p-steuer) [inline usr/lib/common/parser.y:204]: looks like something is wrong with identation here.

File: usr/lib/common/parser.y [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    parsedata = private;
+    
+	do {
+		yyparse();
+
+	} while (!feof(yyin));
+
+	fclose(conf);
+    parsefuncs = NULL;
```
Comment (p-steuer) [inline usr/lib/common/parser.y:213]: and here

**Comment by p-steuer** [comment]
Great piece of work. Any ideas how to test that old config files do not become invalid (language inclusion is generally undecidable even for context free grammars). We dont have fuzzing in place yet. Maybe add a few small well-formed and ill-formed config file examples to the test dir and check for known anwers ? Those could later be used as input hints for the fuzzer.

File: usr/lib/common/parser.y (hunk 1 above)
Comment (juergenchrist) [reply usr/lib/common/parser.y:204]: Unfortunately the file was wrongly detected by my setup and indented with spaces.  Fixed by now.

File: usr/lib/common/parser.y (hunk 2 above)
Comment (juergenchrist) [reply usr/lib/common/parser.y:213]: Fixed.

**Comment by juergenchrist** [comment]
I have some tests mostly around spacing.  Those are some tests that the old parser refused while the new parser allows them.  I was searching for a good tool to fuzz-test this, but I did not find a tool that takes a scanner- and a parser-definition and generates grammar-based input.

I will compile a set of config files that I will then turn into a regression test for the current parser (both, accept and reject cases).

**Review by ifranzki — APPROVED** [review]
lgtm

**Pull request merged**